chore(adr): add adr for reassign evolve (#74)
ar-io/ar-io-ant-process: docs/madr/1-reassign-evolve.md

# Reassign ANT to Evolve Implementation Kickoff on the ArNS Portal

- Status: Proposed
- Approvers: [Ariel], [Dylan], [Atticus], [Phil], [David]
- Date: [2025-01-30]
- Authors: [Atticus]

## Context and Problem Statement

Currently to evolve the ANT we use AOS `Eval` - this introduces the problem of
evaluating what the current version of the code is. If we maintain a variable in
the state of the process containing a version number, that tells us only what
that version number is. At any point the user - owner - of the ANT may `Eval`
new code into the process resulting in changes to the process that may not
reflect what we expect the version number we assigned to reflect. To overcome
this we dryrun interactions on APIs and analyze the results to see if the APIs
perform as expected - this is very inefficient as dry run calls to CU's are
expensive in networking client side, as well as for the CU.

> And so it is proposed to leverage the `Reassign-Name` handler in conjunction

Reviewer: this is the output, not necessarily the contet/problem.

The main problem is upgrading ANTs doesn't guarantee users have the necessary APIs and specs to interact with the arns app. Additionally gateways expect ANTs to properly return names/comply with API specs and if they do not then arns name resolution is impacted

Author: [17f5a95](https://github.com/ar-io/ar-io-ant-process/pull/74/commits/17f5a959051231ca92bf0469ae591c9b9c25e887)

with `_boot` to fork the process state to a whole new ANT, assign that new ANT
as the registered ANT to the specified ArNS name in the registry, and using the
module ID of the process to identify process version and capabilities, and

> solely the module ID.

Reviewer: `With the ARIO contract now support `Reassign-Name` and the recent additions to add `_boot` handler...`

add add links referencing these changes (the PRs are fine)

Author: [e55f7a1](https://github.com/ar-io/ar-io-ant-process/pull/74/commits/e55f7a1f9a963d263bf50cf267d8e2dc28d3921d)

## Decision Drivers

> - Improve the performance of ANT loading client side.

Reviewer: no mention of the benefit of modules > evals

Author: [264df77](https://github.com/ar-io/ar-io-ant-process/pull/74/commits/264df77c0c653b2b9cb2fa9fa4f630fb06b8a6e4)

- Ensure the upgrade process is more robust by using immutable module IDs.
- Reduce computational load on CU infra.
- Mitigate risks associated with reassigning ANT state and ARNS records.

## Considered Options

1. **Current Approach (Baseline)**
   - Multiple dry runs per ANT.
   - ETH compatibility checked via handler analysis.
   - Upgrade involves loading Lua code.

> 2. **Optimized Approach (Proposed Change)**

Reviewer: ```suggestion
2. **Leverage Re-Assign Name and _boot capabilites to upgrade ANTs**
```

Author: [c2fa070](https://github.com/ar-io/ar-io-ant-process/pull/74/commits/c2fa070068bb4e2624552e5a5bbb992828b8c079)

   - Use the module ID to evaluate process version.
   - Upgrade via spawning a new ANT and reassigning ARNS records.

## Decision Outcome

### Chosen Option: Optimized Approach

The optimized approach reduces computational overhead and improves caching for
better performance. It also shifts to a more robust upgrade mechanism using
immutable module IDs rather than version detection heuristics. We can do this
now that we compile and publish our own WASM binaries, rather than using an AOS
binary and loading our code into it.

### Positive Consequences

> - **Performance Improvement:** Reducing dry runs significantly speeds up ANT

Reviewer: gateways can choose to support certain ANTs based on their module ID

Author: [9ec2e7a](https://github.com/ar-io/ar-io-ant-process/pull/74/commits/9ec2e7a81774fbe2108f2e391238ffa0564d58de)

  loading.
- **Immutable Module Versioning:** Eliminates issues with heuristic version
  analysis.

> - **Reduced CU Load:** Fewer computations per ANT.

Reviewer: add some deatils about how the module versions can be used by downstream clients to validate what capabilities and ANT has

Author: [cb34240](https://github.com/ar-io/ar-io-ant-process/pull/74/commits/cb3424057f52aae133ed45c362cc1668a6aca977)

   - ETH compatibility checked via handler analysis.
   - Upgrade involves loading Lua code.
2. **Leverage Re-Assign Name and \_boot capabilites to upgrade ANTs**
   - Use the module ID to evaluate process version and supported capabilites
   - Upgrade via spawning a new ANT and reassigning ARNS records.

## Decision Outcome

### Chosen Option: **Leverage Re-Assign Name and \_boot capabilites to upgrade ANTs**

With the ARIO process now supporting
[`Reassign-Name`](https://github.com/ar-io/ar-io-ant-process/pull/26) and recent
additions to add [`_boot`](https://github.com/ar-io/ar-io-ant-process/pull/57)
to fork the process state to a whole new ANT, we can assign that new ANT as the
registered ANT to the specified ArNS name in the registry, and using the module
ID of the process to identify process version and capabilities, and solely the
module ID.

This optimized approach reduces computational overhead and improves caching for
better performance. It also shifts to a more robust upgrade mechanism using
immutable module IDs rather than version detection heuristics. We can do this
now that we compile and publish our own WASM binaries, rather than using an AOS

> binary and loading our code into it.

Reviewer: Additionally, we can store the latest module ids in a process - enabling support to push new ANT modules dynamically, as they are produced (e.g. publish new module, update the repository process, and arns app pulls latest modules from process to inform user of upgrades). 

TLDR - I'd like to understand more what we get from this approach

Author: [3a4a3c0](https://github.com/ar-io/ar-io-ant-process/pull/74/commits/3a4a3c09e94e2f4827b2e5ac787629224ef80fd8)

```mermaid
sequenceDiagram
    autonumber

    participant ARIOGateway
    participant Owner
    participant OldAnt
    participant NewAnt
    participant AntRegistry
    participant ARIOProcess

    Owner ->> ARIOGateway: Get Module ID for process from GQL
    ARIOGateway -->> Owner: Module ID Result

    alt No upgrade needed Module ID from process meta is up to date

        else

        Owner->>OldAnt: Get state
        OldAnt-->>Owner: Return state

        Owner->>NewAnt: Spawn new ant with old state
        NewAnt->>AntRegistry: Register ANT notice
        Owner->>NewAnt: Get State to verify responsiveness and Module ID to verify correct deployment

        Owner->>OldAnt: Sign and accept Reassign message
        OldAnt->>ARIOProcess: Forward Reassign Message
        ARIOProcess-->>OldAnt: Reassign-Notice

        Owner->>ARIOProcess: Check for Reassignment success and notify user

    end
```

### Positive Consequences

- **Performance Improvement:** Reducing dry runs significantly speeds up ANT
  loading.

> - **Immutable Module Versioning:** Eliminates issues with heuristic version

Reviewer: we discussed this and i do not see it,

by publishing a module and validating it's capabilities before reassinging a name, we can guarantee the new ANT works as expected AND has the latest features/fixes as described by the module. We could not do this before with evals

Author: I actually had it further down but added it here as well [59241af](https://github.com/ar-io/ar-io-ant-process/pull/74/commits/59241affbdf49e4f55853e7c811acf02db2343ab)

  analysis.
  - Downstream clients can identify and validate the capabilities of a module
    (and by extension the ANT itself) has by maintaining a map of modules that
    are capables of executing target workflows
- **Reduced CU Load:** Fewer computations per ANT.
- **Gateway ANT Module Whitelisting:** Gateways can decide which ANT Modules
  they support.

### Negative Consequences

- **Complicates Memory Usage on ANT Registry:** More ANTs means more state in
  the ANT registry, which currently doesn't have a cleanup process.
- **Potential Integration Issues:** Existing integrations relying on fixed ANT
  IDs may break.
- **State Limitations:** Large ANT states (e.g., >2000 undernames) may fail to
  bootstrap on spawn, meaning we need to understand the limitations there and
  decide what amounts we wish to support.
- **MU cranking risks**: Its possible a module fails to instantiate/boot
  properly - the defense there is to query for its state to ensure

>   responsiveness

Reviewer: once spawned. Users will not be presented to reassign a name if spawning fails, or a newly created process is not as expected.

Author: [e8606a0](https://github.com/ar-io/ar-io-ant-process/pull/74/commits/e8606a049e2e40d833911437d53af117306c13d9)

- **Custom ANTs will be lost**: If a user has a custom ANT implementation
  (meaning code that is outside of our control is loaded in) then this code and
  state will not be carried over.

## Pros and Cons of the Options

### Current Approach

- `+` Already implemented and functional.
- `+` Ensures workflow compatibility via handler checks and dryruns.
- `-` Multiple dry runs (in our case 3) increases computational cost for CUs.
- `-` Upgrade process relies on heuristic version checks.

> - `-` No way to assert utility of the ANT without the dryruns.

Reviewer: ```suggestion
- `-` No way to assert capabiliities and schema compliance of the ANT without multiple dryruns.
```

Author: [b4e263e](https://github.com/ar-io/ar-io-ant-process/pull/74/commits/b4e263e95aae76a900580966bc9dd3dcb09fcf48)

- `-` Relying on Eval poses a security risk.

### **Leverage Re-Assign Name and \_boot capabilites to upgrade ANTs** (Chosen)

> - `+` 3x faster ANT loading.

Reviewer: please clarify

  IDs may break.
- **State Limitations:** Large ANT states (e.g., >2000 undernames) may fail to
  bootstrap on spawn, meaning we need to understand the limitations there and
  decide what amounts we wish to support.
- **MU cranking risks**: Its possible a module fails to instantiate/boot
  properly - the defense there is to query for its state to ensure
  responsiveness once spawned. Users will not be presented to reassign a name if
  spawning fails, or a newly created process is not as responding as expected.
- **Custom ANTs will be lost**: If a user has a custom ANT implementation
  (meaning code that is outside of our control is loaded in) then this code and
  state will not be carried over.

## Pros and Cons of the Options

### Current Approach

- `+` Already implemented and functional.
- `+` Ensures workflow compatibility via handler checks and dryruns.
- `-` Multiple dry runs (in our case 3) increases computational cost for CUs.
- `-` Upgrade process relies on heuristic version checks.
- `-` No way to assert capabilities and schema compliance of the ANT without
  multiple dryruns.
- `-` Relying on Eval poses a security risk.

> 

Reviewer: `- Evals don't allow to validate behavior of the ANT after eval occurs`

Author: https://github.com/ar-io/ar-io-ant-process/pull/74/commits/4c64200eda9969fab608f83f83370f8088d93d64

### **Leverage Re-Assign Name and \_boot capabilites to upgrade ANTs** (Chosen)

- `+` 3x faster ANT loading due to no longer needing to perform heuristic
  dryruns for version checking.
- `+` Reduces CU consumption.
- `+` Enables robust versioning.

>   - We can store modules in a version registry allowing clientside applications

Reviewer: add
`+` Ability to validate updated ANT works before reassigning name

Author: one point down we have ```
- `+` Gives us the ability to guarantee that we will not brick the ANT, because
  we can spawn it then validate it, unlike with Eval where we cannot eval first,
  then test after.
```
Want me to still add this or edit this point

# Reassigning ANTs as Evolve/Upgrade Mechanism

> - Status: Proposed

Reviewer: ```suggestion
- Status: Accepted
```
